### PR TITLE
[Spec] Implement n-gram speculative decoding proposer

### PR DESCRIPTION
**Pull request by eplatero97**
# N-gram Speculative Decoding Proposer

## Description

Implements n-gram speculative decoding as a third `MetalProposer` strategy for vLLM Metal. N-gram matching enables efficient prefix prediction by scanning the KV cache for matching token sequences and proposing candidate tokens for speculation.

## Changes

- **`vllm_metal/v1/ngram_proposer.py`** — New `NgramProposer` class implementing n-gram prefix matching:
  - Configurable n-gram size and proposal depth
  - Scans KV cache for token sequence matches
  - Generates deterministic candidate proposals without model inference

- **`vllm_metal/v1/spec_decode.py`** — Integrated `NgramProposer` into the speculative decoding pipeline:
  - Updated `MetalProposer` factory to support n-gram mode
  - Routing logic to select proposer based on configuration

- **`vllm_metal/v1/model_runner.py`** — Model runner updates to support n-gram proposer initialization and invocation

- **`docs/speculative_decoding.md`** — Documentation updated to describe n-gram speculative decoding technique and configuration

## Testing

- **`tests/test_ngram_spec_decode.py`** — Unit tests validating:
  - N-gram matching correctness
  - Candidate selection logic
  - Edge cases (short sequences, no matches, etc.)

- **`tools/test_ngram_spec_decode_e2e.py`** — End-to-end integration test:
  - Full inference pipeline with n-gram proposer
  - Verification against reference implementation

All tests pass locally. Run with:
```bash
pytest -m "not slow" tests/test_ngram_spec_decode.py -v
scripts/test.sh  # Full CI suite
```

## Related

This work builds on the existing speculative decoding infrastructure and complements the draft model proposer (#447).

### CODE REVIEW

File: vllm_metal/v1/spec_decode.py [hunk 1]
```diff
@@ -94,7 +94,11 @@ def validate_supported(
         if use_async_scheduling and (
```
Comment (WindChimeRan) [inline vllm_metal/v1/spec_decode.py]: can be simplified to `if use_async_scheduling and speculative_config is not None`

File: vllm_metal/v1/ngram_proposer.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        # active and num_tokens_no_spec is the committed history length.
+        num_requests = len(drafting)
+        num_tokens_no_spec = np.array(
+            [len(state.token_ids) for _, state in drafting], dtype=np.int32
+        )
+        width = max(int(num_tokens_no_spec.max()), 1)
+        token_ids_cpu = np.zeros((num_requests, width), dtype=np.int32)
+        for i, (_, state) in enumerate(drafting):
+            token_ids_cpu[i, : len(state.token_ids)] = state.token_ids
+        sampled_token_ids: list[list[int]] = [[0]] * num_requests
```
Comment (WindChimeRan) [inline vllm_metal/v1/ngram_proposer.py:130]: [[0] for _ in range(num_requests)]

**Review by WindChimeRan — COMMENTED** [review]
`_collect_drafting_requests()` is a near-verbatim copy of `DraftModelProposer._collect_draft_plans()`

--- 

Not sure if we should address this in the current PR, or a followup refactoring.

File: vllm_metal/v1/ngram_proposer.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            token_ids_cpu,
+        )
+
+        req_ids: list[str] = []
+        draft_token_ids: list[list[int]] = []
+        for (req_id, _), draft in zip(drafting, drafts, strict=True):
+            if not draft:
+                continue
+            req_ids.append(req_id)
+            draft_token_ids.append([int(token) for token in draft])
```
Comment (WindChimeRan) [inline vllm_metal/v1/ngram_proposer.py]: is the int cast necessary?

File: vllm_metal/v1/ngram_proposer.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        # Upstream marks a row "active" by a non-empty sampled-ids entry; the
+        # match itself reads only token_ids_cpu[i, :num_tokens_no_spec[i]]. We
+        # forward exactly the requests we have decided may draft, so every row is
+        # active and num_tokens_no_spec is the committed history length.
+        num_requests = len(drafting)
+        num_tokens_no_spec = np.array(
+            [len(state.token_ids) for _, state in drafting], dtype=np.int32
+        )
+        width = max(int(num_tokens_no_spec.max()), 1)
+        token_ids_cpu = np.zeros((num_requests, width), dtype=np.int32)
```
Comment (WindChimeRan) [inline vllm_metal/v1/ngram_proposer.py]: maybe we can mv token_ids_cpu to __init__? currently each step malloc a token_ids_cpu

**Review by WindChimeRan — COMMENTED** [review]
```
PROMPTS = {
    "getter_setter": (
        "Write Python getter and setter methods for the following fields: "
        "name, age, email, phone, address, city, state, zip_code, country. "
        "Follow this exact pattern for each field:\n\n"
        "    def get_name(self):\n"
        "        return self._name\n\n"
        "    def set_name(self, name):\n"
        "        self._name = name\n\n"
        "Write all getter/setter pairs now:"
    ),
    "echo_numbers": (
        "1 2 3 4 5 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5 "
        "1 2 3 4 5 1 2 3 4 5 1 2 3 4 5 1 2 3 4 5"
    ),
    "json_repeat": (
        'Generate a JSON array of 10 users. Each user has the same structure: '
        '{"name": "...", "age": ..., "email": "..."}. '
        'Here is the first entry: {"name": "Alice", "age": 30, "email": "alice@example.
com"}. '
        "Now generate all 10:"
    ),
    "natural_prose": (
        "Explain the theory of general relativity in simple terms."
    ),
}
```

Local test MODEL = "Qwen/Qwen3-0.6B" on the prompt above. 
It works. Results:

| Prompt            | Drafted | Accepted | Rate      |
| ----------------- | ------- | -------- | --------- |
| **echo_numbers**  | 150     | 150      | **100%**  |
| **natural_prose** | 180     | 118      | **65.6%** |
| **json_repeat**   | 147     | 81       | **55.1%** |
| **getter_setter** | 132     | 60       | **45.5%** |
| **Overall**       | **609** | **409**  | **67.2%** |

**Review by WindChimeRan — COMMENTED** [review]
LGTM. Please check the nits. 

Only the `vllm_metal/v1/spec_decode.py` condition check is a clear blocker. Others are small nits. 

Thanks!

**Comment by eplatero97** [comment]
> `_collect_drafting_requests()` is a near-verbatim copy of `DraftModelProposer._collect_draft_plans()`
> 
> Not sure if we should address this in the current PR, or a followup refactoring.

I think we should address it in this PR since it's fresh. I placed shared logic into `SpeculativeDecodeController.draft_eligible_requests()` so that both SpD methods (and future ones) can now use. 

Let me know what you think (I also addressed previous comments, which were good feedback).

both files, `tests/test_ngram_spec_decode.py` and `tools/test_ngram_spec_decode_e2e.py`, are passing.

I got similar results with same prompts but using `max_model_len=1024` due to only having 8GB of RAM (:
```
Expected baseline (indicative, not asserted):

    Prompt          Drafted     Accepted    Rate
    getter_setter   ~150        ~117        ~78%
    echo_numbers    ~150        ~150        ~100%
    json_repeat     ~147        ~81         ~55%
    natural_prose   ~180        ~118        ~66%
    Overall         ~627        ~466        ~74%
```

**Review by WindChimeRan — APPROVED** [review]
